File: frontend/src/types/crm.ts
export type Etapa =
  | "a_discovery"
  | "discovery_agendada"
  | "en_diagnostico"
  | "presentacion_agendada"
  | "en_cierre"
  | "ganado"
  | "descartado"

export type Oportunidad = {
  id: string
  nombre: string
  email?: string
  telefono?: string
  empresa?: string
  tamano?: string
  ubicacion?: string
  fuente?: string
  etapa: Etapa
  vendedor_id?: string
  producto_recomendado?: string
  observaciones?: string
  created_by?: string
  created_at: string
  updated_at: string
  etapa_changed_at?: string
  eliminado: boolean
}

export type Actividad = {
  id: string
  oportunidad_id: string
  autor_id?: string
  tipo: string
  nota?: string
  etapa_anterior?: string
  etapa_nueva?: string
  created_at: string
}

export type Archivo = {
  id: string
  oportunidad_id: string
  nombre: string
  descripcion?: string
  tipo: string
  storage_path: string
  subido_por?: string
  tamano_bytes?: number
  created_at: string
}

export const ETAPA_LABEL: Record<Etapa, string> = {
  a_discovery: "A Discovery",
  discovery_agendada: "Discovery Agendada",
  en_diagnostico: "En Diagnóstico",
  presentacion_agendada: "Presentación Agendada",
  en_cierre: "En Cierre",
  ganado: "Ganado",
  descartado: "Descartado",
}

export const ETAPAS: Etapa[] = [
  "a_discovery",
  "discovery_agendada",
  "en_diagnostico",
  "presentacion_agendada",
  "en_cierre",
  "ganado",
  "descartado",
]

File: backend/app/schemas/crm.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, EmailStr

from app.models.oportunidad import CrmEtapa


class OportunidadCreate(BaseModel):
    nombre: str
    email: Optional[EmailStr] = None
    telefono: Optional[str] = None
    empresa: Optional[str] = None
    tamano: Optional[str] = None
    ubicacion: Optional[str] = None
    fuente: Optional[str] = None
    etapa: CrmEtapa = CrmEtapa.a_discovery
    vendedor_id: Optional[UUID] = None
    producto_recomendado: Optional[str] = None
    observaciones: Optional[str] = None


class OportunidadUpdate(BaseModel):
    nombre: Optional[str] = None
    email: Optional[EmailStr] = None
    telefono: Optional[str] = None
    empresa: Optional[str] = None
    tamano: Optional[str] = None
    ubicacion: Optional[str] = None
    fuente: Optional[str] = None
    vendedor_id: Optional[UUID] = None
    producto_recomendado: Optional[str] = None
    observaciones: Optional[str] = None


class EtapaUpdate(BaseModel):
    etapa: CrmEtapa


class EliminarRequest(BaseModel):
    razon: str


class OportunidadOut(BaseModel):
    id: UUID
    nombre: str
    email: Optional[str] = None
    telefono: Optional[str] = None
    empresa: Optional[str] = None
    tamano: Optional[str] = None
    ubicacion: Optional[str] = None
    fuente: Optional[str] = None
    etapa: CrmEtapa
    vendedor_id: Optional[UUID] = None
    producto_recomendado: Optional[str] = None
    observaciones: Optional[str] = None
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime
    etapa_changed_at: Optional[datetime] = None
    eliminado: bool

    model_config = {"from_attributes": True}


class ActividadCreate(BaseModel):
    tipo: str  # nota | llamada | email
    nota: Optional[str] = None


class ActividadOut(BaseModel):
    id: UUID
    oportunidad_id: UUID
    autor_id: Optional[UUID] = None
    tipo: str
    nota: Optional[str] = None
    etapa_anterior: Optional[str] = None
    etapa_nueva: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ArchivoCreate(BaseModel):
    nombre: str
    descripcion: Optional[str] = None
    tipo: str
    storage_path: str
    tamano_bytes: Optional[int] = None


class ArchivoOut(BaseModel):
    id: UUID
    oportunidad_id: UUID
    nombre: str
    descripcion: Optional[str] = None
    tipo: str
    storage_path: str
    subido_por: Optional[UUID] = None
    tamano_bytes: Optional[int] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class VendedorOut(BaseModel):
    id: UUID
    email: str
    role: str
    is_active: bool

    model_config = {"from_attributes": True}
